src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -2213,6 +2251,83 @@ METHOD(kernel_ipsec_t, destroy, void,
 	free(this);
 }
 
+struct tunnel_counter_key {
+	uint32_t terminator_id;
+	const char* tenant_id;
+	const char* account_id;
+	const char* vrf_name;
+};
+
+static bool tunnel_counter_key_compare(const void *key, const void *other_key) {
+    struct tunnel_counter_key *tunnel_1 = (struct tunnel_counter_key *)key;
+    struct tunnel_counter_key *tunnel_2 = (struct tunnel_counter_key *)other_key;
+
+	return tunnel_1->terminator_id == tunnel_2->terminator_id &&
+			!strcmp(tunnel_1->tenant_id, tunnel_2->tenant_id) && 
+			!strcmp(tunnel_1->account_id, tunnel_2->account_id) && 
+			!strcmp(tunnel_1->vrf_name, tunnel_2->vrf_name);
+}
+
+static u_int tunnel_counter_key_hash(const void *key) {
+	struct tunnel_counter_key *tunnel_key = (struct tunnel_counter_key *)key;
+	char tunnel_key_str[MAX_ACCOUNT_ID_LEN + MAX_TENANT_ID_LEN + MAX_VRF_NAME_LEN + 10];
+	sprintf(tunnel_key_str, "%u%s%s%s", tunnel_key->terminator_id, tunnel_key->tenant_id, tunnel_key->account_id, tunnel_key->vrf_name);
+	return hashtable_hash_str(&tunnel_key_str);
+}
+
+static bool count_tunnels(private_kernel_super_ipsec_t* this, struct pm_terminator_conn* const terminator_conn) {
+	const struct pm_proto_addr lookup_addr = *picomsg_terminator_conn_get_lookup_addr(terminator_conn);
+	uint32_t term_id = picomsg_terminator_conn_get_id(terminator_conn);
+	enumerator_t* enumerator;
+	ike_sa_t* ike_sa;
+	uint64_t sa_counter = 0;
+	hashtable_t *counter_map = NULL;
+
+	counter_map = hashtable_create(tunnel_counter_key_hash, tunnel_counter_key_compare, 10);
+
+	enumerator = charon->controller->create_ike_sa_enumerator(charon->controller, true);
+	while (enumerator->enumerate(enumerator, &ike_sa)) {
+		struct pm_proto_addr ike_sa_addr;
+		host_addr_to_picomsg_addr(ike_sa->get_my_host(ike_sa), &ike_sa_addr);
+		if (pm_proto_addr_equals(&ike_sa_addr, &lookup_addr)) {
+			struct super_cert_info cert_info;
```

Comment:
It is more readable to have as few nestings as possible in the code. This means instead of writing the core logic inside the "good branch" of the `if` statement, change the `if` to have an early exit:
``` cpp
while (enumerator->enumerate(enumerator, &ike_sa)) {
    if (! pm_proto_addr_equals(...)) {
        continue;
    }
    if (! parse_peer_certificate(...)) {
        continue;
    }

    const struct tunnel_counter_key curr_key = ...
}
```
(some coding styles even **mandate** structuring the code like this)